### PR TITLE
Bit exact gaussian blur for 16bit unsigned int

### PR DESCRIPTION
**Pull request by Yosshi999**
In the current implementation, bit-exact gaussian blur supports only 8-bit unsigned int.
To create more precise blurred images with bit-exactness, it is necessary to support fixedpoint matrix.
In this PR you can create blurred images by using `ufixedpoint16`.
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/imgproc/src/fixedpoint.inl.hpp [hunk 1]
```diff
@@ -157,18 +157,21 @@ class ufixedpoint64
     CV_ALWAYS_INLINE bool isZero() { return val == 0; }
     static CV_ALWAYS_INLINE ufixedpoint64 zero() { return ufixedpoint64(); }
     static CV_ALWAYS_INLINE ufixedpoint64 one() { return ufixedpoint64((uint64_t)(1ULL << fixedShift)); }
+
+    static CV_ALWAYS_INLINE ufixedpoint64 fromRaw(uint64_t v) { return ufixedpoint64(v); }
+    CV_ALWAYS_INLINE ufixedpoint64 raw() { return val; }
```
Comment (terfendail) [inline modules/imgproc/src/fixedpoint.inl.hpp:162]: I think return type should be `uint64_t`

File: modules/imgproc/src/smooth.dispatch.cpp [hunk 2]
```diff
@@ -258,23 +258,20 @@ softdouble getGaussianKernelFixedPoint_ED(CV_OUT std::vector<int64_t>& result, c
...
     std::vector<int64_t> fixed_256;
-    softdouble approx_err = getGaussianKernelFixedPoint_ED(fixed_256, res_sd, 8);
+    softdouble approx_err = getGaussianKernelFixedPoint_ED(fixed_256, res_sd, FT::fixedShift);
     CV_UNUSED(approx_err);
 
     res.resize(n);
     for (int i = 0; i < n; i++)
     {
-        res[i] = ufixedpoint16::fromRaw((uint16_t)fixed_256[i]);
+        res[i] = FT::fromRaw(fixed_256[i]);
```
Comment (terfendail) [inline modules/imgproc/src/smooth.dispatch.cpp:274]: Implicit type casting here cause warnings on Win builds. I think there are two options to handle that:
1. Extend fixedpoint classes with typedef of native internal type to make possible explicit type cast here
2. Update fromRaw definition to receive int64_t as input (or add one more fromRaw with such input parameter)

File: modules/imgproc/src/smooth.simd.hpp [hunk 3]
```diff
@@ -54,9 +54,10 @@
 namespace cv {
 CV_CPU_OPTIMIZATION_NAMESPACE_BEGIN
 // forward declarations
+template <typename RFT>
 void GaussianBlurFixedPoint(const Mat& src, /*const*/ Mat& dst,
-                            const uint16_t/*ufixedpoint16*/* fkx, int fkx_size,
-                            const uint16_t/*ufixedpoint16*/* fky, int fky_size,
+                            const RFT/*FT*/* fkx, int fkx_size,
```
Comment (terfendail) [inline modules/imgproc/src/smooth.simd.hpp:59]: Could you please remove comments here

**Review by terfendail — CHANGES_REQUESTED** [review]

File: modules/imgproc/src/smooth.dispatch.cpp (hunk 2 above)
Comment (Yosshi999) [reply modules/imgproc/src/smooth.dispatch.cpp]: I chose 1 (FT::raw_t)

File: modules/imgproc/src/fixedpoint.inl.hpp [hunk 4]
```diff
@@ -159,7 +161,7 @@ class ufixedpoint64
     static CV_ALWAYS_INLINE ufixedpoint64 one() { return ufixedpoint64((uint64_t)(1ULL << fixedShift)); }
 
     static CV_ALWAYS_INLINE ufixedpoint64 fromRaw(uint64_t v) { return ufixedpoint64(v); }
-    CV_ALWAYS_INLINE ufixedpoint64 raw() { return val; }
+    CV_ALWAYS_INLINE uint64_t raw() { return val; }
```
Comment (terfendail) [inline modules/imgproc/src/fixedpoint.inl.hpp:164]: Could you please update `CV_ALWAYS_INLINE ufixedpoint32 raw() { return val; }` and `CV_ALWAYS_INLINE ufixedpoint16 raw() { return val; }` as well

File: modules/imgproc/src/fixedpoint.inl.hpp [hunk 5]
```diff
@@ -261,19 +268,23 @@ class ufixedpoint32
     CV_ALWAYS_INLINE bool isZero() { return val == 0; }
     static CV_ALWAYS_INLINE ufixedpoint32 zero() { return ufixedpoint32(); }
     static CV_ALWAYS_INLINE ufixedpoint32 one() { return ufixedpoint32((1U << fixedShift)); }
+
+    static CV_ALWAYS_INLINE ufixedpoint32 fromRaw(uint32_t v) { return ufixedpoint32(v); }
+    CV_ALWAYS_INLINE ufixedpoint32 raw() { return val; }
```
Comment (terfendail) [inline modules/imgproc/src/fixedpoint.inl.hpp:273]: Could you please update return type here and for `CV_ALWAYS_INLINE ufixedpoint16 raw()` as well

**Review by terfendail — APPROVED** [review]

**Comment by terfendail** [comment]
@Yosshi999 Could you please rebase to the latest 3.4 branch

**Pull request merged**